// roots/lib/cli/tpl.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Roots, antimatter, help;

  antimatter = require('anti-matter');

  Roots = require('../');

  module.exports = function(args, cli) {
    var promise;
    args = args._;
    if (args.length < 2) {
      return cli.emit('data', help());
    }
    promise = (function() {
      switch (args[1]) {
        case 'add':
          return Roots.template.add({
            name: args[2],
            uri: args[3]
          });
        case 'remove':
          return Roots.template.remove({
            name: args[2]
          });
        case 'list':
          return cli.emit('data', Roots.template.list({
            pretty: true
          }));
        case 'default':
          return Roots.template["default"](args[2]);
        case 'reset':
          return Roots.template.reset();
      }
    })();
    if (promise.then) {
      return promise.done(function(r) {
        return cli.emit('data', r.green);
      }, function(e) {
        return cli.emit('err', ("" + e).red);
      });
    }
  };

  help = function() {
    return antimatter({
      title: 'Roots Templates',
      options: {
        width: 65,
        color: 'blue'
      },
      commands: [
        {
          name: 'add',
          required: ['name', 'url'],
          description: 'Adds a new roots template called [name], located at [url], which is any url that can be "git clone"-d'
        }, {
          name: 'remove',
          required: ['name'],
          description: 'Removed the template with the given [name]'
        }, {
          name: 'list',
          description: 'Lists all installed roots templates'
        }, {
          name: 'default',
          required: ['name'],
          description: 'Makes the template [name] the default template whenever `roots new` is run'
        }
      ]
    });
  };

}).call(this);
